gt/main.c

```diff
@@ -870,14 +794,310 @@ print_unsent_policy(struct ggu_policy *policy)
+		sizeof(struct udp_hdr) + sizeof(struct ggu_common_hdr);
+	struct ggu_decision *ggu_decision = rte_pktmbuf_mtod_offset(
+		ggu_pkt->buf, struct ggu_decision *, offset);
+	unsigned int decision_list_len = ggu_pkt->buf->data_len - offset;
+	ggu_policy_iterator(ggu_decision, decision_list_len,
+		print_unsent_policy, NULL, "gt");
+}
+
+/*
+ * Find the notification packet being buffered for the Gatekeeper
+ * server specified in @pkt_info, if any.
+ */
+static struct ggu_notify_pkt *
+find_notify_pkt(struct gt_packet_headers *pkt_info,
+	struct gt_instance *instance)
+{
+	unsigned int i;
+
+	if (instance->num_ggu_pkts == 0)
+		return NULL;
+
+	for (i = 0; i < instance->max_ggu_pkts; i++) {
+		struct ggu_notify_pkt *ggu_pkt = &instance->ggu_pkts[i];
+
+		if (ggu_pkt->buf == NULL)
+			continue;
+
+		if (pkt_info->outer_ethertype != ggu_pkt->ipaddr.proto)
+			continue;
+
+		if (ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv4) {
+			struct ipv4_hdr *ipv4_hdr =
+				(struct ipv4_hdr *)pkt_info->outer_l3_hdr;
+			if (ggu_pkt->ipaddr.ip.v4.s_addr ==
+					ipv4_hdr->src_addr)
+				return ggu_pkt;
+		} else if (likely(ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv6)) {
+			struct ipv6_hdr *ipv6_hdr =
+				(struct ipv6_hdr *)pkt_info->outer_l3_hdr;
+			if (ipv6_addrs_equal(ipv6_hdr->src_addr,
+					ggu_pkt->ipaddr.ip.v6.s6_addr))
+				return ggu_pkt;
+		}
+	}
+	return NULL;
+}
+
+static void
+send_notify_pkt(struct gt_instance *instance, struct ggu_notify_pkt *ggu_pkt)
+{
+	struct udp_hdr *notify_udp;
+
+	/*
+	 * Complete the packet fields that can only be done
+	 * when the packet is ready to be transmitted.
+	 */
+
+	if (ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv4) {
+		struct ipv4_hdr *notify_ipv4 =
+			rte_pktmbuf_mtod_offset(ggu_pkt->buf,
+				struct ipv4_hdr *, ggu_pkt->buf->l2_len);
+		notify_ipv4->total_length = rte_cpu_to_be_16(
+			ggu_pkt->buf->data_len - ggu_pkt->buf->l2_len);
+
+		/* Offload the UDP checksum. */
+		notify_udp = (struct udp_hdr *)&notify_ipv4[1];
+		notify_udp->dgram_cksum = rte_ipv4_phdr_cksum(notify_ipv4,
+			ggu_pkt->buf->ol_flags);
+	} else if (likely(ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv6)) {
+		struct ipv6_hdr *notify_ipv6 =
+			rte_pktmbuf_mtod_offset(ggu_pkt->buf,
+				struct ipv6_hdr *, ggu_pkt->buf->l2_len);
+		notify_ipv6->payload_len = rte_cpu_to_be_16(
+			ggu_pkt->buf->data_len - ggu_pkt->buf->l2_len -
+			sizeof(struct ipv6_hdr));
+
+		/* Offload the UDP checksum. */
+		notify_udp = (struct udp_hdr *)&notify_ipv6[1];
+		notify_udp->dgram_cksum = rte_ipv6_phdr_cksum(notify_ipv6,
+			ggu_pkt->buf->ol_flags);
+	} else {
+		rte_panic("Unexpected condition: gt at lcore %u sending notification packet to Gatekeeper server with unknown IP version %hu\n",
+			rte_lcore_id(), ggu_pkt->ipaddr.proto);
+	}
+
+	notify_udp->dgram_len =
+		rte_cpu_to_be_16((uint16_t)(ggu_pkt->buf->data_len -
+			ggu_pkt->buf->l2_len - ggu_pkt->buf->l3_len));
+
+	if (rte_eth_tx_burst(get_net_conf()->front.id, instance->tx_queue,
+			&ggu_pkt->buf, 1) != 1) {
```

Comment:
While it's okay, it's odd that we use `get_net_conf()` instead of adding `gt_config` to the parameters of `send_notify_pkt()`. The way I think of it is that the XXX_config structs define a namespace, as a container in the Linux kernel. So calling `get_net_conf()` here is potentially violating this namespace. Of course, for now, it doesn't make a difference in Gatekeeper.

The goal of this patch is to **batch** policy decisions, but we are sending single packets out instead of batching them into the array of packets to go out at the end of processing a batch of incoming packets in `gt_proc()`.